File: dynamic_programming/kth_permutation.rb
# https://leetcode.com/problems/permutation-sequence/

def kth_permutation(n, k) # 4, 16
  permutation = []
  elements    = (1..n).to_a # [1, 2, 3, 4]

  # number of permutations for each position
  factorials  = [nil, 1] # undefined for less than 1 character
  (2..n).map do |idx|
    # if n > 1, figure out how many permutations exist
    factorials << idx * factorials[idx.pred]
  end

  k = k.pred # permutations begin at index #0
  until n == 0
    # use previous entry in factorials array if n > 1
    # otherwise divide current permutations by n
    part_length = factorials[n.pred] || factorials[n] / n
    part_idx = k / part_length

    # permutation starts with element at part_idx
    # discard element after detrermining its position in permutation
    permutation << elements.delete_at(part_idx)

    n = n.pred
    k %= part_length
  end
  permutation.join.to_i
end

puts kth_permutation(4, 16) == 3241
puts kth_permutation(3, 3)  == 213
puts kth_permutation(4, 9)  == 2314
puts kth_permutation(3, 1)  == 123

File: dynamic_programming/kth_permutation.py
def factorial(n):
  if n == 0: 
    return 1
  return n * factorial(n - 1)

def getPermutation(n: int, k: int) -> str:
  def kth_perm(string, k): #[1,2,3], 2
    permutation = [] 
    length = len(string)
    fact = factorial(length)
    while (string != []): 
      fact = fact // length
      index, k = divmod(k, fact)
      x = string[index] 
      permutation.append(x) 
      string = string[:index] + string[index+1:] 
      length -= 1
    return "".join(permutation)
  return kth_perm([str(i) for i in range(1, n+1)], k-1)

def kth_permutation(n, k):
  permutation = []
  unused =  list(range(1, n + 1))
  factorial = [1] * (n + 1)
  for idx in range(1, n+1):
    factorial[idx] = idx * factorial[idx - 1]
  k -= 1
  while n > 0:
    part_length = factorial[n] // n
    idx = k // part_length
    permutation.append(unused[idx])
    unused.pop(idx)
    n -= 1
    k %= part_length
  return ''.join(map(str, permutation))

print(getPermutation(3,3)) # 213
print(getPermutation(4,9)) # 2314
print(getPermutation(3,1)) # 123

print(kth_permutation(3,3)) # 213
print(kth_permutation(4,9)) # 2314
print(kth_permutation(3,1)) # 123

